### DISCUSSION TITLE
Custom Server - Pass props/context to pages & layouts on render()

### DISCUSSION THREAD

**Discussion by yanivfranco** [vercel/next.js#50168] [Ideas]
Hey guys, for a long time now, I am facing issues with custom server and SSR. I hoped app router will fix these issues, but I see its not the case. Maybe someone here could help me out... and add this functionality🙏 

Let's say I use dynamic layout theme data which I want to pass to my main layout component with my custom server.
With the `pages` directory I could use the `query` object as a hacky way to pass this data as follows:

In the server (using the `query` param to render):
```
const query = {foo: "bar"};
server.render(req, res, '/', query);
```

Component:
```
const Home: NextPage = (props: any) => {
  return <Theme foo={props.foo}>
    <h1>Welcome</h1>
  </Theme>;
};

export async function getServerSideProps(context: NextPageContext) {
  return {
    props: {
      foo: context.query.foo,
    },
  };
}

export default Home;
```

This implementation is not ideal as getServerSideProps runs for each page instead of rendering the layout once on app start. (Also, lack of getServerSideProps on `_app` is a big issue)

With the new `appRouter` I really hoped this issue will be fixed, only to find out that its not, and furthermore, I can't use my hacky solution anymore because now `layout` components does not receive the `searchParams` object.

I wish the `render` function had some `context` or `props` param that will be passed to the server components & layouts. Is there some functionality I am missing in order to help me with this issue? I am getting frustrated...
What do you think?

Thanks

**Comment by poorscousertommy8** [comment]
I have exactly the same request. How can I submit data to pages.js in custom server?

Were you able to find a solution?
